libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -75,39 +148,40 @@ void DeclarationTypeChecker::endVisit(UserDefinedTypeName const& _typeName)
 		fatalTypeError(_typeName.location(), "Name has to refer to a struct, enum or contract.");
 	}
 }
-void DeclarationTypeChecker::endVisit(FunctionTypeName const& _typeName)
+bool DeclarationTypeChecker::visit(FunctionTypeName const& _typeName)
 {
+	if (_typeName.annotation().type)
+		return false;
+
+	bool previousInsideFunctionType = m_insideFunctionType;
+	m_insideFunctionType = true;
```

Comment:
Can you explain what this variable is used for? I'm not completely sure what the semantics are

Reply:
```
struct S {
  function() returns (S memory) f;
}
```
The inner ``S`` occurs inside the function type and does not constitute an actually recursive use of S (at least not a problematic one).

Reply:
Ah ok